lib/preflight/preflighttest.js

```diff
@@ -0,0 +1,483 @@
+    ]).then(([networkQuality, collectedStats]) => {
+      collectedStats.networkQuality = networkQuality;
+      return collectedStats;
+    });
+  }
+
+  /**
+   * returns a promise to executes given step
+   * rejects the return promise if
+   * a) preflight is stopped.
+   * b) subscriber or publisher disconnects
+   * c) step does not complete in reasonable time.
+   * @param {function} step - function to execute
+   * @param {string} stepName - name for the step
+   */
+  function executePreflightStep(stepName, step) {
+    const MAX_STEP_DURATION = testDuration + 10 * SECOND;
+    if (preflightTest._stopped) {
+      throw new Error('stopped');
+    }
+
+    if (subscriberRoom && subscriberRoom.state === 'disconnected') {
+      throw new Error('subscriber disconnected unexpectedly');
+    }
+
+    if (publisherRoom && publisherRoom.state === 'disconnected') {
+      throw new Error('publisher disconnected unexpectedly');
+    }
+
+    const stepPromise = Promise.resolve().then(step);
+    let timer = null;
+    const timeoutPromise = new Promise((_resolve, reject) => {
+      timer = setTimeout(() => {
+        reject(new Error(`Timed out waiting for : ${stepName}`));
+      }, MAX_STEP_DURATION);
+    });
+    return Promise.race([timeoutPromise, stepPromise]).finally(() => {
+      clearTimeout(timer);
+    });
+  }
+
+  return Promise.resolve()
+    .then(() => {
+      return executePreflightStep('acquire media', () => {
+        return [createAudioTrack(), createVideoTrack()];
+      });
+    }).then(tracks => {
+      return executePreflightStep('connect publisher', () => {
+        localTracks = tracks;
+        preflightTest.emit('progress', PreflightProgress.mediaAcquired);
+        return connect(publisherToken, options);
+      });
+    }).then(room => {
+      return executePreflightStep('connect subscriber', () => {
+        publisherRoom = room;
+        connectTiming = new TimeMeasurement();
+        mediaTiming = new TimeMeasurement();
+        return connect(subscriberToken, Object.assign(options, { name: room.sid }));
+      });
+    }).then(room => {
+      return executePreflightStep('publisher sees subscriber', () => {
+        subscriberRoom = room;
+        connectTiming.stop();
+        preflightTest.emit('progress', PreflightProgress.connected);
+        trackStartListener = listenForTrackStarted(subscriberRoom, 2);
+        return participantsConnected(publisherRoom);
+      });
+    }).then(() => {
+      return executePreflightStep('subscriber sees publisher', () => {
+        return participantsConnected(publisherRoom);
+      });
+    }).then(() => {
+      return executePreflightStep('validate rooms', () => {
+        if (subscriberRoom.sid !== publisherRoom.sid) {
+          throw new Error('wrong room tokens. Ensure that both token join to the same room');
+        }
+        if (subscriberRoom.participants.size !== 1) {
+          throw new Error('unexpected participant found in the room');
```

Comment:
How about `'Unexpected participant found in the room. Ensure that the room has no participants in it.'`?